### PR TITLE
src: find_first_of-related perf improvements

### PR DESCRIPTION
**Pull request by miguelteixeiraa**
refs: https://github.com/ada-url/ada/issues/219
This PR is not the result of an analysis of all uses of **find_first_of** in the repo. The perf gain is not that big, so it might be interesting to consider things like code readability before merging. 

```
**This PR**
time: 2776 ns
instructions/ns=18.4029 
instructions/url=4.60073k

**Without this PR**
time: 2834 ns
instructions/ns=18.61
instructions/url=4.72355k
```

### CODE REVIEW

File: src/helpers.cpp [hunk 1]
```diff
@@ -94,26 +94,35 @@ namespace ada::helpers {
...
+          // Is the URL valid???
+          break;
+        }
+      } else if (url.is_special()) {
+        if (view[location] == ':' || view[location] == '/' || view[location] == '?' || view[location] == '\\') {
+          break;
+        }
       } else {
-        location = url.is_special() ? view.find_first_of(":[/?\\#", location) : view.find_first_of(":[/?#", location);
+        if (view[location] == ':' || view[location] == '/' || view[location] == '?') {
```
Comment (anonrig) [inline src/helpers.cpp:115]: We can remove this extra branch by checking is special and the condition that comes with it

File: src/helpers.cpp [hunk 2]
```diff
@@ -94,26 +94,35 @@ namespace ada::helpers {
...
-    // Next while loop is almost never taken!
-    while((location != std::string_view::npos) && (view[location] == '[')) {
-      location = view.find(']',location);
-      if(location == std::string_view::npos) {
-        inside_brackets = true;
-        /**
-         * TODO: Ok. So if we arrive here then view has an unclosed [,
-         * Is the URL valid???
-         */
+ada_really_inline size_t get_host_delimiter_location(const ada::url& url, std::string_view& view, bool& inside_brackets) noexcept {
```
Comment (lemire) [inline src/helpers.cpp:97]: Instead of passing `const add:url & url`, we can make the function conceptually simpler by passing `const bool is_special`.

**Comment by lemire** [comment]
Great work. I managed to improve it a bit at ...

https://github.com/ada-url/ada/pull/232

But @miguelteixeiraa has achieved 90% of the gains I got already.

We should continue: we can get more gains still by hacking similar code pieces.

**Comment by miguelteixeiraa** [comment]
> Great work. I managed to improve it a bit at ...
> 
> #232
> 
> But @miguelteixeiraa has achieved 90% of the gains I got already.
> 
> We should continue: we can get more gains still by hacking similar code pieces.

Amazing!

**Comment by miguelteixeiraa** [comment]
Let's close this one and merge https://github.com/ada-url/ada/pull/232 !

**Comment by anonrig** [comment]
Thanks @miguelteixeiraa. Can you also review @lemire's pull request?

**Pull request closed without merging**